### DISCUSSION TITLE
Rotor tilt provided at runtime

### DISCUSSION THREAD

**Discussion by misi9170** [NatLabRockies/floris#653] [General]
This discussion is concerned with whether we should allow users to provide turbine tilt angles at runtime (for example, in the `calculate_wake` method of `FlorisInterface`) in a manner similar to how yaw angles are provided, or whether tilt angles should be part of the turbine definition only (the status quo).

Generally speaking, turbine tilt angles are not controllable in the way that yaw angles are, and the current implementation of rotor tilt allows (uncontrollable) specification of the tilt angle in the turbine yaml using:
- `ref_tilt_cp_ct` (presumably, the design tilt angle of the turbine rotor)
- `floating_tilt_table` (for floating turbines, specifies the tilt angle at different wind speeds)

However, there may be users who would like to quickly vary the tilt angles external to the turbine definition file. Allowing methods on `FlorisInterface` to accept tilt angles in a manner similar to yaw angles would provide this. Code to do part of this job is already in place in `calculate_wake()`, but has been commented out (#651) because it is not complete. This code would need to be propagated to other methods on `FlorisInterface` if we decide to move forward with allowing tilt angles to be specified at runtime.

If we decide to allow users to specify tilt angles at run time, we will need to decide how to reconcile the provided tilt angle with the value provided from the `floating_tilt_table`, if the turbine is floating. Presumably, the provided tilt angle should override the value from the floating tilt table, but this would need to be implemented and verified. Further, we would need to decide whether a user-provided tilt angle is taken as a delta from `ref_tilt_cp_ct` or as an absolute value---this was not an issue for the yaw angle, as we implicitly assumed that the reference yaw angle is 0 (or at least, 0 mean).

**Comment by misi9170** [comment]
@bayc @paulf81 Hoping to get your thoughts here when you have the time. If we decide to provide this sort of functionality, it would be after the imminent release.

**Reply by clown991** [reply]
When I use it, I change the tilt angle of the wind turbine by directly changing the tilt value in the YAML file. However, when there are multiple wind turbines, either the angle is fixed, or each wind turbine has different customs that can be changed through the tilt table. I cannot specify each one independently. After reading this discussion, I really hope to improve this function as soon as possible. In addition, can you provide an interface to know the current status of each wind turbine after each run? Thank you very much. Looking forward to your reply.

**Reply by clown991** [reply]
And when I was in AWC mode and changed the angle of the 3-tilt at the same time, I found that the power of the first wind turbine remained unchanged

**Reply by misi9170** [reply]
Hi @clown991 , we do not currently have the capability of actively setting the tilt angle of individual turbines to arbitrary values and will not likely have this any time in the near future.

If you would like to implement this yourself, you could consider a couple of options after you have cloned the FLORIS code so that you have a local copy:
1. The most thorough approach would be to mimic the way that `yaw_angles` are passed through the code from the top level at [`FlorisModel.set()`](https://github.com/NREL/floris/blob/fd0e599ef01dc9f1d426d2b4c99e2f6beb57fefd/floris/floris_model.py#L373) through to where they are used in the [solver](https://github.com/NREL/floris/blob/fd0e599ef01dc9f1d426d2b4c99e2f6beb57fefd/floris/core/solver.py).
2. As an alternative "hack", you could pass the tilt angles in to the `yaw_angles` argument at the top level and then overwrite the tilt angles with the passed in `yaw_angles` inside the [solver](https://github.com/NREL/floris/blob/fd0e599ef01dc9f1d426d2b4c99e2f6beb57fefd/floris/core/solver.py). This will only work if you do not want to apply both tilt and yaw angles.
3. Another simpler (but possibly less elegant) hack would be to just hard code the tilt angles you want to test at the top of [solver](https://github.com/NREL/floris/blob/fd0e599ef01dc9f1d426d2b4c99e2f6beb57fefd/floris/core/solver.py). This could be the easiest option if you only want to check a few different tilt angles.

Note that only the [Empirical Gaussian model](https://nrel.github.io/floris/empirical_gauss_model.html) supports vertical wake deflections due to tilt, and these have not been validated for correctness. See [this line of the `empirical_guass_solver()`](https://github.com/NREL/floris/blob/fd0e599ef01dc9f1d426d2b4c99e2f6beb57fefd/floris/core/solver.py#L1321). The other models will still model the effect of tilt on rotor quantities such as thrust coefficient, axial induction, and power, but will not cause the wake to deflect in the vertical direction.

Regarding your question about the AWC mode: the [AWC operation model](https://github.com/NREL/floris/blob/fd0e599ef01dc9f1d426d2b4c99e2f6beb57fefd/floris/core/turbine/operation_models.py#L510) does not currently support nonzero tilt or yaw angles, because we do not (yet) have a good understanding of the combined effect of blade-pitch actuated wake mixing and rotor misalignment. Therefore, any tilt (or yaw) angles provided to the AWC operation model will be ignored, and you'll receive the same power output.

**Reply by clown991** [reply]
Thank you very much for your detailed reply！
